contracts/poap/src/contract.rs

```diff
@@ -0,0 +1,451 @@
+    // Check that the poap uri is a valid IPFS url
+    let poap_uri = Url::parse(&msg.event_info.base_poap_uri)
+        .map_err(|_err| ContractError::InvalidPoapUri {})?;
+    if poap_uri.scheme() != "ipfs" {
+        return Err(ContractError::InvalidPoapUri {});
+    }
+
+    // Check that the event uri is a valid IPFS url
+    let event_uri =
+        Url::parse(&msg.event_info.event_uri).map_err(|_err| ContractError::InvalidEventUri {})?;
+    if event_uri.scheme() != "ipfs" {
+        return Err(ContractError::InvalidEventUri {});
+    }
+
+    // Check per address limit
+    if msg.event_info.per_address_limit == 0 {
+        return Err(ContractError::InvalidPerAddressLimit {});
+    }
+
+    let config = Config {
+        admin: admin.clone(),
+        minter: minter.clone(),
+        per_address_limit: msg.event_info.per_address_limit,
+        cw721_code_id: msg.cw721_code_id,
+        mint_enabled: false,
+    };
+    // Save the received event info.
+    CONFIG.save(deps.storage, &config)?;
+
+    let event_info = EventInfo {
+        creator: creator.clone(),
+        start_time: msg.event_info.start_time,
+        end_time: msg.event_info.end_time,
+        base_poap_uri: msg.event_info.base_poap_uri.clone(),
+        event_uri: msg.event_info.event_uri.clone(),
+    };
+    // Save the event info
+    EVENT_INFO.save(deps.storage, &event_info)?;
+
+    // Submessage to instantiate cw721 contract
+    let sub_msgs: Vec<SubMsg> = vec![SubMsg {
+        msg: WasmMsg::Instantiate {
+            admin: Some(admin.to_string()),
+            code_id: msg.cw721_code_id,
+            msg: to_binary(&Cw721InstantiateMsg {
+                name: msg.cw721_initiate_msg.name,
+                symbol: msg.cw721_initiate_msg.symbol,
+                minter: env.contract.address.to_string(),
+            })?,
+            funds: info.funds,
+            label: "poap cw721".to_string(),
+        }
+        .into(),
+        id: INSTANTIATE_CW721_REPLY_ID,
+        gas_limit: None,
+        reply_on: ReplyOn::Success,
+    }];
+
+    Ok(Response::new()
+        .add_attribute("method", "instantiate")
+        .add_attribute("creator", creator)
+        .add_attribute("admin", admin)
+        .add_attribute("minter", minter)
+        .add_attribute("start_time", msg.event_info.start_time.to_string())
+        .add_attribute("end_time", msg.event_info.end_time.to_string())
+        .add_attribute(
+            "per_address_limit",
+            msg.event_info.per_address_limit.to_string(),
+        )
+        .add_attribute("base_poap_uri", &msg.event_info.base_poap_uri)
+        .add_attribute("event_uri", &msg.event_info.event_uri)
+        .add_attribute("cw721_code_id", &msg.cw721_code_id.to_string())
+        .add_submessages(sub_msgs))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
```

Comment:
Should we have a function `msg.validate()` before executing?  What do you think @manu0466, @bragaz?

Reply:
Thinking about having the validation logic in a single place seems reasonable to me, we can have this methd that checks that the received data is correct but i prefer to have the permissions checks inside the execution function to make explicit that the authorization checks where properly performed.

Reply:
Yeah, permissions checks only can be performed inside the execution process, `msg.validate` is for checking if the value of msg is valid, say, the check of `end_time` must be after `start_time` inside the msg can be validated here.
With this function, the execute function can be focused on verifying `permission` or `state` except for the msg data.

Reply:
I agree with the introduction of the `validate()` function for the msg fields checks.